lib/core/engine/iteration.js

```diff
@@ -61,6 +61,57 @@ class Iteration {
     this.asyncScriptsByCategory = asyncScriptsByCategory;
   }
 
+  async runPrePostScript(script, driver, context, commands) {
+    let requiresPrivilege = false;
+
+    if (typeof script === 'object') {
+      const scriptObject = script;
+      if (
+        typeof scriptObject.function !== 'function' ||
+        typeof scriptObject.requires !== 'object'
+      ) {
+        log.error(
+          'New-style pre/post script detected ' +
+            'but with invalid construction. ' +
+            'It will not execute.'
+        );
+        script = null;
+      } else {
+        log.info('Working with a new-style script.');
+        script = scriptObject.function;
+        requiresPrivilege =
+          scriptObject.requires.privilege === true ? true : false;
+      }
+    }
+    if (
+      requiresPrivilege &&
+      (typeof driver.getContext !== 'function' ||
+        typeof driver.setContext !== 'function')
+    ) {
+      log.error(
+        'Only browsers with get/setContext functions ' +
+          'can run privileged JavaScript.'
+      );
+      script = null;
+    }
+
+    if (script !== null) {
+      let oldContext = null;
+      if (requiresPrivilege) {
+        oldContext = driver.getContext();
+        log.info('Setting privileged context!');
+        await driver.setContext('chrome');
+      }
+
+      await script(context, commands);
+
+      if (oldContext !== null) {
```

Comment:
I started to say: "Explain to me why you need to manage contexts yourself here.  Why can't you use the existing API for privileged execution (https://github.com/mozilla/browsertime/blob/9d7e3715744a065fd69afd0681ee1b192f119876/lib/core/seleniumRunner.js#L295), etc."

But in fact all of this pre/post stuff is already possible, i.e.: https://github.com/bdekoz/browsertime/blob/privileged-telemetry/vendor-scripts/clear-persistent-data-stop.js.

So there are two issues: first, what is the goal of these pre/post changes?  Second, the context is only relevant on the target device, and pre/post scripts run on the host machine.  So I don't quite understand how that's going to work.

Can you separate the pre/post stuff into a part 2 commit that builds on the main privileged measurement commit?  That will be the simplest way for us to isolate the needed pieces.

Reply:
> I started to say: "Explain to me why you need to manage contexts yourself here. Why can't you use the existing API for privileged execution (
> 
> https://github.com/mozilla/browsertime/blob/9d7e3715744a065fd69afd0681ee1b192f119876/lib/core/seleniumRunner.js#L295
> ), etc."
> 
> But in fact all of this pre/post stuff is already possible, i.e.: https://github.com/bdekoz/browsertime/blob/privileged-telemetry/vendor-scripts/clear-persistent-data-stop.js.
> 
> So there are two issues: first, what is the goal of these pre/post changes? Second, the context is only relevant on the target device, and pre/post scripts run on the host machine. So I don't quite understand how that's going to work.

This is technically true. *But*, the pre/post scripts get a `context` that can be indirected to execute a script on the target browser. In the example(s) that I posted, that's exactly how I set the collection mask:

```
let run = () => {
  /*
   * Tell the engine that we want to know about all the possible
   * performance stats that are available!
   */
  ChromeUtils.setPerfStatsCollectionMask(255);
};
await context.selenium.driver.executeScript(run).then(value => {});
```

For that to work, the target browser would have to be "primed" with the privileges required. That, at least, was the way I thought about it. 

I will separate out this change into a subsequent PR against the other changes in this PR once this PR lands ... wow, that's a mouthful! That way, as you said, we can work together to think about the mechanics of if/how this works!

Is that okay?

> 
> Can you separate the pre/post stuff into a part 2 commit that builds on the main privileged measurement commit? That will be the simplest way for us to isolate the needed pieces.